accepted/2024/trusted-publishers-oidc-for-nuget-push.md

```diff
@@ -0,0 +1,568 @@
+experience where API keys (and therefore trust policies) are not set on the organization. Instead, they are set by the
+user who is a member of the organization. The package will be recorded as being pushed by a specific user *on behalf of*
+an organization instead of being pushed by the organization directly. 
+
+As with the current experience with long-lived API keys, the API key is limited in what it can do based on the scopes
+assigned to the trust policy.
+
+If a symbol package (.snupkg) is produced along with the .nupkg, it can also be pushed using the same API key. `dotnet
+nuget push` handles this automatically.
+
+### Step 3: execute the workflow to push the package to NuGet.org
+
+In the sample above, the workflow has a `workflow_dispatch` trigger meaning it can be queued manually. However your
+workflow is triggered, it will perform the following steps as described in the workflow YAML.
+
+A flow chart of data flow between the Trusted Publisher (Identity Provider) and the Package Repository (NuGet.org) is
+shown below. This is lovingly *stolen* from the OpenSSF Trusted Publishers document linked above.
+
+![Trusted Publishers flow](meta/trusted-publishers-oidc-for-nuget-push/trusted-publishers-flow.png)
+
+This describes a two-legged authentication flow. The first leg is trading an OIDC token for an API key. The second leg
+is using the API key for a privileged action, like package upload.
+
+In written form, this happens:
+
+1. The `nuget/login` workflow step does this:
+   - Fetch a GitHub Actions OIDC token, enabled via the `id-token: write` workflow permission
+   - Send the OIDC token to a NuGet.org endpoint, along with the NuGet.org username
+1. NuGet.org does this:
+   - Receive the OIDC token, verify the token using the public GitHub Actions keyset and other JWT rules
+   - Verify the provided username has a trust policy matching the OIDC token
+   - Generate a short-lived API key and return it to the caller
+1. The `nuget/login` workflow step does this:
+   - Receive the short-lived API key and set it as a step output
+2. The `dotnet nuget push` workflow step does this:
+   - Use the provided API key to perform a package push, using the [`PackagePublish`
+     resource](https://learn.microsoft.com/en-us/nuget/api/package-publish-resource)
+3. NuGet.org does this:
+   - Verify the provided package and package operation match the scopes on the API key
+   - Publish the package!
+
+## Technical explanation
+
+See the [technical document](trusted-publishers-oidc-for-nuget-push.technical.md) in this case directory for more
+information.
+
+## Drawbacks
+
+This workflow only helps users who want to build and push their packages on a recognized Trusted Publisher. Not all
+CI/CD systems will be supported so some users will still need to use the older authentication flows such as long-lived,
+manually rotated API keys.
+
+The trust policy is configured once and works forever. An API key forces the package author to re-assess CI/CD
```

Comment:
Hmm good question! I think starting out we would allow full flexibility in a "walk-crawl-run" approach. As we mature, we see how the feature goes in our ecosystem, we can lock things down more. We took a similar approach with requiring 2FA on our website. At first it was opt in, then opt out (new accounts had it enabled), then it was mandated. This allowed us to look at customer feedback as we went and not move so quickly that folks got stuck.

Reply:
I added a section about this.